src/position.rs

```diff
@@ -0,0 +1,144 @@
+    let mut bam = threaded_bam_reader(input_path, nthreads);
+    let header = bam::Header::from_template(bam.header());
+    let bam_bytes = fs::metadata(&input_path).unwrap().len();
+    let buckets = ((bam_bytes / max_mem) + 1) as u32;
+
+    let output = if input_path == output_path {
+        let mut hasher = DefaultHasher::new();
+        hasher.write(output_path.as_bytes());
+        format!("{}.{}.tmp", output_path, hasher.finish())
+    } else {
+        output_path.to_string()
+    };
+    
+    info! {"Processing file {} of size {} into {} buckets", input_path, bam_bytes, buckets };    
+    clean_buckets(); 
+    bucketify_bam(&mut bam, &header, buckets);
+    process_buckets(&header, &output, buckets, nthreads);
+    clean_buckets(); 
+
+    if output != output_path {
+        fs::remove_file(output_path).unwrap();
+        fs::rename(output, output_path).unwrap();
+    }
+}
+
+fn bucketify_bam(bam: &mut bam::Reader, header: &bam::Header, nbuckets: u32) {
+    fs::create_dir_all(Path::new("buckets")).unwrap();
+    let mut writers: Vec<_> = bucket_names("buckets", nbuckets).iter()
+        .map(|p| bam::Writer::from_path(p, &header, bam::Format::Bam).unwrap())
+        .collect();
+    
+    for r in bam.records() {
+        let record = match r {
+            Ok(r) => {
+                trace!("Processing record: {:?}", r);
+                r
+            }
+            Err(e) => {
+                info! { "failed to read BAM record {:?}", e };
+                continue
+            }
+        };
+        //Grab the read name and figure out which bucket it goes into
+        let qname = record.qname();
+        let mut hasher = DefaultHasher::new();
+        hasher.write(qname);
+        let hash = hasher.finish();
+        let bucket = (hash % (nbuckets as u64)) as usize;
+        //XXX: Not sure if this is buffered I/O or not, if so, we'll have to manually buffer
+        writers[bucket].write(&record).expect("Failed to write record");
+    }
+}
+
+fn process_buckets(header: &bam::Header, output_path: &str, nbuckets: u32, nthreads: usize) {
+    let output_path = Path::new(output_path);
+    fs::create_dir_all(output_path.parent().unwrap()).expect("failed to create output directories");
+    if output_path.exists() {
+        fs::remove_file(output_path).expect("output file exists and cannot be removed")
+    }
+    let mut writer = bam::Writer::from_path(output_path, &header, bam::Format::Bam).unwrap();
+
+    let mut readers: Vec<_> = (0..nbuckets)
+        .map(|x| threaded_bam_reader(&format!("buckets/bucket_{}.bam", x), nthreads))
+        .collect();
+    
+    for reader in &mut readers {
+        process_bucket(reader, &mut writer)
+    }
+}
+
+fn process_bucket(reader: &mut bam::Reader, writer: &mut bam::Writer) {
+    //XXX: avoid re-hashing these records every time
+    let mut records: Vec<_> = reader.records().filter_map(|r| r.ok()).collect();
```

Comment:
Not sure about the memory layout of this vector. If it is a vector containing the records themselves (instead of pointers to them), and it is presumably a large vector, it might be worth trying to speed-up things a bit:
* we should know the size of every bucket that were created. Hence we could initialize the vector with a given size and then populate it (to avoid the size doubling / copy operations)
* sort a vector of indices into that record vector to avoid the copying cost during sorting.

Reply:
Yes, this stores the record struct itself and not a pointer to it, but the record struct is actually a fairly thin wrapper around the data, which is stored as a `*mut u8` pointer in the struct:

https://docs.rs/rust-htslib/latest/rust_htslib/htslib/struct.bam1_t.html

There's another iteration method on the bam::Reader type, rc_records(), that gives you an iterator over references, `Rc<Record>`, instead of just Record. The indirection actually seems to make things about 5% slower, though. (A 1.5 GB file went took ~2:35min instead of ~2:20.

With the record sizes being relatively light, I would be extraordinarily surprised if the other overhead working on indexes instead of just the records directly would be an improvement. I can check, but I think there's probably low-hanging fruit elsewhere?